### PR TITLE
[RFC] Merge wm4 master

### PR DESCRIPTION
**Pull request by xantoz**
I'm putting this here as a reference so we can see and discuss about how/what we want to include this/what can be backported etc. or just for curious people to play with the backwards playback :)

This of course includes some of the more controversial commits such as the one removing support for DVB, DVD, bluray etc. (like discussed in #6127) so that might be a good starting point. Personally I feel that the path of least resistance at this point would be to investigate the amount of work needed to bring wanted removed features back on top of this branch, due to how much the demuxing code has changed since.

### CODE REVIEW

**Comment by olifre** [comment]
I tried to have a go at bringing back DVB support on top of this branch. 
In most cases, it is pretty straightforward, but the channel switching works via properties, which goes via `prop_stream_ctrl` which uses `demux_stream_control` which is gone in this branch. 

I think something similar is true for most other removed features. Bringing them back (on top of this branch) is probably only possible when also bringing back `demux_stream_control`, and all the related stuff, like `demux_control` etc.

**Comment by bjin** [comment]
my -1 vote on merging "backward playback" changes. Tried it and I believe it's only "useful" to demonstrate how creepy backward played audio is. However, the code change is huge and buggy (from [commit message](https://github.com/mpv-player/mpv/pull/6781/commits/2b39a48448efb68fbe10d9042540c39923eea205)). Assuming @wm4 won't interact with this repo directly, it could be maintainance nightmare.

**Comment by xantoz** [comment]
@bjin Yeah, I see where you're coming from, although I think the state of the backwards playback improves a bit after that particular commit.

Re: maintainability, some of the changes to the demuxer is to make it easier to maintain, which is nice. I don't know how well one could separate those changes from the backwards playback hacks, though.

**Comment by olifre** [comment]
> Re: maintainability, some of the changes to the demuxer is to make it easier to maintain, which is nice.

I checked the most heavy changes to the demuxer code, and a major part (just in lines of code) was exactly the removal of the stream control code. So the improved maintainability goes hand in hand with dropping all the mentioned features, and is (to a large part) from the reduced functionality. 
A part of the removed functionality has been brought back in separated functions - adding those back needed by the removed features would probably be more messy than the previous code (since that would bring back the commands handled via `enum`s before all back as separate functions).

**Comment by CounterPillow** [comment]
If we're keeping DVD playback, we should probably exclude 7ce243010412c9c075301b25f1764adeef6bdc2d and 6073e99c093d619f99e25949f07973f8ec69d28f 65a1c90b33700d7226c4f4929a4bb201306bae09 and 8f4844b35c46ac6c2828f38d8ff8f4ffaf8e96e1

Though I personally don't care about keeping DVD playback, but some users might still have actual optical drives in their computers that they use.

**Comment by Argon-** [comment]
Considering the high output of wm4, I'm not sure if it's a good idea to even think about invasive changes (or changes at all) to the commits to be merged. It just doesn't work well in the long run unless someone pledges to maintain these changes here and really sticks to it. Not sure if this is very reasonable.

Removing DVD/BD support is really not gonna be fun with users but honestly, a) how many people are still using these technologies and b) are using them with mpv, considering there are better working alternatives out there? And does someone want to pledge a serious time commitment to this group?

**Comment by olifre** [comment]
> Removing DVD/BD support is really not gonna be fun with users but honestly, a) how many people are still using these technologies and b) are using them with mpv, considering there are better working alternatives out there?

Are there better working alternatives, with comparable playback quality? Please mind that this also relates to playing back DVDs stored on HDD. 
Also, these commits remove DVB and TV. There are alternatives out there, but (at least for DVB) these are either full-fledged media centers, or half-assed solutions (c.f. VLC) which break in day-to-day usage. No well working media player with consistent tuning interface and support for all DVB broadcasting techniques comes to mind (I have contributed some things which were missing to mpv exactly because I found no good alternative, and have spent some significant time there, and I have also spent quite some timing handling all user's bug reports until now finally it seems we are stable for any corner case). 

But that's a discussion we already had in #6127 . I would even be willing to bring back DVB on top of this branch, but it does not seem very feasible, since it essentially means to revert many of the demuxer changes.

**Comment by olifre** [comment]
FWIW: I could bring back DVB support (not TV support!) easily, on top of this branch, without _any_ changes to the demuxer code - but loosing the channel switching functionality, and the properties exposing the channel name, number and frequency (and allowing to toggle them at runtime). 

If it's really decided we need these commits merged, this is the route I would go. I personally do not use the channel switching often, but use the recording and re-encoding functionality mostly, and start the player with the channel I want to watch, so I'd personally be fine with it. 

But this may make some people who spent development time on optimizing the channel switching, such as @rozhuk-im , rather unhappy (and of course - users). 
If there's interest, I can clean up the necessary commit and push it somewhere, I readied and tested it just now, it just needs some more polishing (documentation mainly).

**Comment by xantoz** [comment]
Re: TV, that can actually still be made to work via av:// after the removal, although with less features probably (none that I really need, but I only ever use it for composite/S-video capture devices).

I think CD as well, although with some slight issues (as covered in #6127). DVD to an extent (probably less features)

**Comment by FranciscoPombal** [comment]
Apologies for the potentially stupid question, but would these commits also remove the capability of playing BD/DVD folder structures/ISOs?

From a user perspective, I would personally prefer it if mpv kept the capability of playing BD/DVD structures/ISOs, and eventually supported menus and other advanced functionality.

**Comment by macdavis** [comment]
I still watch Blu-Ray/DVDs a lot so I do hope this function will not be removed in the future.

I don’t quite understand the code itself, but from the discussion, the problem seems to be the trade off between backward play and some existing functions. I’m not against adding new functions, but just don’t think it’s a good idea to drop existing functions that some users still use for a function that practically almost no one would use.

**Comment by CounterPillow** [comment]
>and eventually supported menus and other advanced functionality.

lol no, what little support for those mpv used to have was removed for a good reason. It made the playback loop code awful, from what I've gathered.

**Comment by xantoz** [comment]
> I don’t quite understand the code itself, but from the discussion, the problem seems to be the trade off between backward play and some existing functions. I’m not against adding new functions, but just don’t think it’s a good idea to drop existing functions that some users still use for a function that practically almost no one would use.

Backward play is really the least significant thing here (but it's fun to play around with). What's more significant is refactoring of a lot of the demuxer code, improvements to handling of DASH, HLS and other  streaming stuff etc. and just lots of misc fixes here and there.

Going by @quot27's analysis in #6127, 
Bluray playback is already supported in FFmpeg, so it could probably be added on quite easily on top of this just by utilising that. The same goes for CD support, and TV support to an extent (not DVB though as I understand it).

DVD would probably take more work to add back on top of this, but I think it could be done more or less (maybe somewhat less featureful, but some basic sort of playback).

**Comment by olifre** [comment]
> not DVB though as I understand it

Exactly - there have been proposals to add DVB to ffmpeg, but it's still not in (yet), and all those proposals had their own issues. 

I have started to bring back DVB support on top of this branch, here (don't pull this yet!):
https://github.com/olifre/mpv/commits/merge-wm4-master-add-dvb
There are basically three parts:
1. Bringing back DVB itself. This is the first commit, and touches almost nothing from demuxer and streamer code. 
2. Bringing back channel switching. I still have to test this with the hardware tonight, but it's not very invasive - it adds back two properties, and wires up the logic again to have them controllable from the player. 
3. Bringing back support for switching between programs in Transport Streams (be it from TS files or from full DVB transponders). Yes, this was removed, too. I'm still working on that. It can be done in a cleaner way than what was there before. 

I think I can bring back everything without crufting demux or stream, while for part 3, I would refactor the logic of the old code a bit to match the new code structure better (that's not committed yet). 

@xantoz : When I am finished with this work and have it actually tested, should I just open a PR against your branch and you could keep those commits on top of this PR? Then we would have the major issue of this series tackled which ffmpeg can not cover yet, while getting the benefits of all the other stuff in this branch :wink:.

**Comment by xantoz** [comment]
> When I am finished with this work and have it actually tested, should I just open a PR against your branch and you could keep those commits on top of this PR? Then we would have the major issue of this series tackled which ffmpeg can not cover yet, while getting the benefits of all the other stuff in this branch

Sounds good to me.

**Comment by olifre** [comment]
> Sounds good to me.

Thanks! It's now open at:
https://github.com/xantoz/mpv/pull/1
All DVB functionality is back and worked fine in my testing. I could refactor more, but for the moment, I left some parts in place which others might want to use to bring other functionality back - the stream controls and demux controls (which had also been removed in the very first commit alongside all those features) are very generic and easily extensible. I brought back the parts of them which were not replaced by more specific functionality. 

TL;DR: All DVB is back. If nothing else is brought back, I could do some more refactoring later.

**Comment by xantoz** [comment]
Cleaned up (only one merge commit) and rebased on latest origin/master. 
Merged the DVB re-additions by @olifre

**Comment by macdavis** [comment]
> Going by @quot27's analysis in #6127,
Bluray playback is already supported in FFmpeg, so it could probably be added on quite easily on top of this just by utilising that.

The point of using libbluray is to get metadata (e.g. Audio/Subtitle Language, Title information) and chapter information, which is still not supported by FFmpeg. Otherwise, for most blu-rays, we can just simply go to /BDMV/STREAM and play the largest file.

**Comment by xantoz** [comment]
Hmm seems like something in your patches, @olifre, is causing it to segfault in open_given_type in demux.c because stream is NULL when playing videos using youtube-dl. At first I thought it was something bad about the rebase I made, but I can reproduce this on olifre/mpv@3db66029287555017a7ea4503d53f7af62b96768 as well.

I can maybe take a deeper look when I have more time (it's getting late)

**Comment by olifre** [comment]
@xantoz Thanks for reporting! https://github.com/xantoz/mpv/pull/2 should fix it. 
My bad was that I tested only with "regular" files and DVB, and missed in the commit series that `stream` may now be `NULL` in `open_given_type` due to changed behaviour of the demuxer.

**Comment by xantoz** [comment]
Rebased on latest origin/master again so we can have one merge commit only. 
Also incorporated the SIGSEGV fix above + restored the old caching decision behavior again (thanks for the help @sfan5)

**Comment by sfan5** [comment]
commits we may not want to merge:
* Remove any pretense for big endian support
  https://github.com/mpv-player/mpv/pull/6781/commits/0d40788299efddca28546a52ade4a2bab71b7480
* stats.lua: silence annoying fontconfig warnings 
  https://github.com/mpv-player/mpv/pull/6781/commits/73c6c6813e4c3a9ab9c08dce4e4001da12612d6d
* Implement backwards playback (+ baggage of >25 commits)
  https://github.com/mpv-player/mpv/pull/6781/commits/2b39a48448efb68fbe10d9042540c39923eea205

how to replace removed functionality:
* cdda: `mpv av://libcdio:/dev/sr0 --start=0`
  - requires start=0 workaround
  - full duration correctly detected, has chapters
* dvd: <b>no replacement</b>
* bluray: `mpv lavf://bluray:path/to/disk/or/bdmv`
  - missing chapters
  - seeking broken in some cases
  - demuxer cache now active (wasn't before)

also, found a regression in caching:
  `mpv ZMCZ-10605.cue`
  cue references a flac in the same folder, which is on a gvfs filesystem
  so it should be cached, but isn't (opening the flac itself caches correctly)

**Comment by qyot27** [comment]
Probably a crazy idea, and one I personally don't have any ability to do, but like with what happened with libplacebo, what if mpv's support for libbluray and dvdnav/dvdread (not including cdda because using that via libavformat is superior, apart from the seek start thing) was split off into a separate library that had a saner, singular API to call into?  That possibly wouldn't require re-inserting stuff that would be removed with this PR, correct?  Not to mention that that could also provide a modern way to get those added into FFmpeg at all (dvd), or as a proper demuxer (bluray).

**Comment by bjin** [comment]
I just noticed that, apparently the backward playback commit also changed the license of `demux.c` to AGPL.This will effectively make both mpv and libmpv as a whole licensed under AGPL, which is a really big change and probably not wanted by some libmpv user.

**Comment by xantoz** [comment]
@bjin well... shit...

**Comment by xantoz** [comment]
I'll try and keep this updated, anyway. It can at least serve as a place to go looking for things that might be useful to backport (as long is it doesn't touch one of the two AGPL:ed files :/)

**Comment by wm4** [comment]
I pushed some of my removal commits.

@olifre ask if you need help how to readd the dvb runtime control

**Comment by olifre** [comment]
@wm4 Many thanks both for the work and keeping DVB in a functional state! I fully agree that dropping the stream_ctrl stuff is a good idea. 
I'm currently slightly overloaded for at least two more weeks, but will get back to add channel switching. 

Do you have an example commit "at hand" which adds an option affecting a stream which can be changed at runtime? This would speed my reading of the mechanism up quite a bit.

**Comment by wm4** [comment]
I couldn't find anything simple. Commit 6aad532aa39481a8100910612fad039dd75236b9 changed quite a lot of OSD-related code from accessing a global option struct to local copies (that are automatically updated in a safe way).

Basically you need to:

1. replace mp_get_config_group with m_config_cache_alloc (actually the former is just a less powerful convenience wrapper for the latter)
2. regularly poll m_config_cache_update (could also use a fancy notification mechanism, but likely not worth doing for this purpose)
3. check which options have been possibly updated

This is uni-directional. Unlike the stream ctrls, it can't give feedback (other than log messages), and can't e.g. reset the global options on failure or anything like it. But this just follows the general way the command interface was changed in the recent years. Instead of making setting properties/option fail if they can't be applied at runtime, there should at most be something that reports the current state, independent of setters.

**Comment by olifre** [comment]
@wm4 Many thanks, while the commit of course is pretty huge, I get the mechanics now. 
I'll either manage to give it a go end of next week or in two weeks from now (too many conferences at the moment which need preparation). 

Uni-directional is perfectly fine in my opinion. On a side-note, I do regard the channel switching as a "bonus feature" of DVB in any case. Switching channels takes noticeably longer (it's actually the DVB hardware's tuning) than restarting mpv, so I doubt many people are "zapping" through channels. 
My personal major usecase is to select a channel and watch it (with filters etc.) or record / reencode it live, so I am rarely using the actual channel switching. Of course I'll spend time bringing channel switching back, there appears to be a user base for it apart from myself and it does not seem too hard now that I got the mechanics. 

Thanks again! :smile:.

**Comment by wm4** [comment]
Obsoleted.
